### PR TITLE
Feat/weight clipper

### LINKED ISSUE #4: [TODO] feature: weight clipper
Python implementation:
https://github.com/open-spaced-repetition/fsrs-optimizer/blob/95694b787bb71ac9883db1201af09e334ee5ee0b/src/fsrs_optimizer/fsrs_optimizer.py#L100-L120

### PR DESCRIPTION
**Pull request by Luc-Mcgrady**
Not sure where to actually call the function but it passes the test I wrote. Hopefully that's what you wanted.
I'm not sure if the method of creating a new tensor is a good one but it seems to run fine.

### CODE REVIEW

File: src/model.rs [hunk 1]
```diff
@@ -93,7 +137,7 @@ impl<B: Backend<FloatElem = f32>> Model<B> {
             );
             let s_forget = self.stability_after_failure(stability, new_d.clone(), r);
             let new_s = s_recall.mask_where(rating.equal_elem(1), s_forget);
-            (new_s.clamp(0.1, 36500.0), new_d)
+            (new_s.clamp(0.1, 36500.0), weight_clipper(new_d))
```
Comment (L-M-Sherlock) [inline src/model.rs:140]: It's not the position to call this function. Weight clipper should be applied to the parameters.

Comment (L-M-Sherlock) [reply src/model.rs:140]: Could you call weight_clipper() in the end of forward()?

**Review by L-M-Sherlock — CHANGES_REQUESTED** [review]

File: src/model.rs (hunk 1 above)
Comment (Luc-Mcgrady) [reply src/model.rs]: Like that?

File: src/model.rs [hunk 2]
```diff
@@ -159,7 +159,7 @@ impl<B: Backend<FloatElem = f32>> Model<B> {
             // dbg!(&difficulty);
             // dbg!()
         }
-        (stability, difficulty)
+        (weight_clipper(stability), weight_clipper(difficulty))
```
Comment (L-M-Sherlock) [inline src/model.rs:162]: The stability and difficulty are not the parameters. I mean `self.w`.

**Review by L-M-Sherlock — CHANGES_REQUESTED** [review]

File: src/model.rs (hunk 2 above)
Comment (L-M-Sherlock) [reply src/model.rs:162]: `pub w: Param<Tensor<B, 1>>` is the parameter.

Comment (Luc-Mcgrady) [reply src/model.rs]: Sorry. like that?

Comment (Luc-Mcgrady) [reply src/model.rs]: hold on nvm

**Comment by L-M-Sherlock** [comment]
Could you run this test?

https://github.com/open-spaced-repetition/fsrs-optimizer-burn/blob/19dc456680d950191f1c929c89eeed46a00e7314/src/training.rs#L132-L146

I get this error:

```
error[E0596]: cannot borrow `*self` as mutable, as it is behind a `&` reference
  --> src/training.rs:34:40
   |
34 |         let (stability, _difficulty) = self.forward(t_historys, r_historys);
   |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable
```

**Comment by Luc-Mcgrady** [comment]
Its because to change self.w I needed to make self mutable and that screwed up some code elsewhere.
Is there any way to change self.w without making self mutable?

**Comment by L-M-Sherlock** [comment]
> Is there any way to change self.w without making self mutable?

I don't know. The maintainer said that they haven't supported it:

https://github.com/burn-rs/burn/issues/662#issuecomment-1684280084

**Comment by Luc-Mcgrady** [comment]
I've given up and made it so that it just clips the weights at the very end.

https://github.com/Luc-Mcgrady/fsrs-optimizer-burn/blob/a6233d9b71ef0eeaacfb1481f716904afcba7593/src/training.rs#L118

I don't think replacing the tensor during training was a good idea anyway.

**Comment by L-M-Sherlock** [comment]
> I don't think replacing the tensor during training was a good idea anyway.

But if we only clips the weights at the end, the errors could be larger than clipping the weights during training.

**Review by L-M-Sherlock — APPROVED** [review]

**Comment by L-M-Sherlock** [comment]
I will merge this branch. It partially implemented the weight clipper.

**Comment by Luc-Mcgrady** [comment]
I meant modifying the w values directly seems like a bad idea, I can see why you would want to clamp it in the forward function.

**Comment by L-M-Sherlock** [comment]
But how could we modify the w values indirectly?

**Pull request merged**